// vite-plugin-compression/dist/index.mjs
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));

// src/index.ts
import path2 from "path";
import { normalizePath } from "vite";

// src/utils.ts
import fs from "fs";
import path from "path";
var isFunction = (arg) => typeof arg === "function";
var isRegExp = (arg) => Object.prototype.toString.call(arg) === "[object RegExp]";
function readAllFile(root, reg) {
  let resultArr = [];
  try {
    if (fs.existsSync(root)) {
      const stat = fs.lstatSync(root);
      if (stat.isDirectory()) {
        const files = fs.readdirSync(root);
        files.forEach(function(file) {
          const t = readAllFile(path.join(root, "/", file), reg);
          resultArr = resultArr.concat(t);
        });
      } else {
        if (reg !== void 0) {
          if (isFunction(reg.test) && reg.test(root)) {
            resultArr.push(root);
          }
        } else {
          resultArr.push(root);
        }
      }
    }
  } catch (error) {
  }
  return resultArr;
}

// src/index.ts
import fs2 from "fs-extra";
import chalk from "chalk";
import zlib from "zlib";
import { debug as Debug } from "debug";
var debug = Debug("vite-plugin-compression");
var extRE = /\.(js|mjs|json|css|html)$/i;
var mtimeCache = new Map();
var exportFn = (options = {}) => {
  let outputPath;
  let config;
  const emptyPlugin = {
    name: "vite:compression"
  };
  const {
    disable = false,
    filter = extRE,
    verbose = true,
    algorithm = "gzip",
    ext = ".gz",
    threshold = 1025,
    compressionOptions = {},
    deleteOriginFile = false,
    success = () => {
    }
  } = options;
  if (disable) {
    return emptyPlugin;
  }
  debug("plugin options:", options);
  return __spreadProps(__spreadValues({}, emptyPlugin), {
    apply: "build",
    enforce: "post",
    configResolved(resolvedConfig) {
      config = resolvedConfig;
      outputPath = path2.isAbsolute(config.build.outDir) ? config.build.outDir : path2.join(config.root, config.build.outDir);
      debug("resolvedConfig:", resolvedConfig);
    },
    async closeBundle() {
      let files = readAllFile(outputPath) || [];
      debug("files:", files);
      if (!files.length)
        return;
      files = filterFiles(files, filter);
      const compressOptions = getCompressionOptions(algorithm, compressionOptions);
      const compressMap = new Map();
      const handles = files.map(async (filePath) => {
        let { mtimeMs, size: oldSize } = await fs2.stat(filePath);
        if (mtimeMs <= (mtimeCache.get(filePath) || 0) || oldSize < threshold)
          return;
        let content = await fs2.readFile(filePath);
        if (deleteOriginFile) {
          fs2.remove(filePath);
        }
        try {
          content = await compress(content, algorithm, compressOptions);
        } catch (error) {
          config.logger.error("compress error:" + filePath);
        }
        const size = content.byteLength;
        const cname = getOutputFileName(filePath, ext);
        compressMap.set(filePath, {
          size: size / 1024,
          oldSize: oldSize / 1024,
          cname
        });
        await fs2.writeFile(cname, content);
        mtimeCache.set(filePath, Date.now());
      });
      Promise.all(handles).then(() => {
        if (verbose) {
          handleOutputLogger(config, compressMap, algorithm);
          success();
        }
      });
    }
  });
};
function filterFiles(files, filter) {
  if (filter) {
    const isRe = isRegExp(filter);
    const isFn = isFunction(filter);
    files = files.filter((file) => {
      if (isRe) {
        return filter.test(file);
      }
      if (isFn) {
        return filter(file);
      }
      return true;
    });
  }
  return files;
}
function getCompressionOptions(algorithm = "", compressionOptions = {}) {
  const defaultOptions = {
    gzip: {
      level: zlib.constants.Z_BEST_COMPRESSION
    },
    deflate: {
      level: zlib.constants.Z_BEST_COMPRESSION
    },
    deflateRaw: {
      level: zlib.constants.Z_BEST_COMPRESSION
    },
    brotliCompress: {
      params: {
        [zlib.constants.BROTLI_PARAM_QUALITY]: zlib.constants.BROTLI_MAX_QUALITY,
        [zlib.constants.BROTLI_PARAM_MODE]: zlib.constants.BROTLI_MODE_TEXT
      }
    }
  };
  return __spreadValues(__spreadValues({}, defaultOptions[algorithm]), compressionOptions);
}
function compress(content, algorithm, options = {}) {
  return new Promise((resolve, reject) => {
    zlib[algorithm](content, options, (err, result) => err ? reject(err) : resolve(result));
  });
}
function getOutputFileName(filepath, ext) {
  const compressExt = ext.startsWith(".") ? ext : `.${ext}`;
  return `${filepath}${compressExt}`;
}
function handleOutputLogger(config, compressMap, algorithm) {
  config.logger.info(`
${chalk.cyan("\u2728 [vite-plugin-compression]:algorithm=" + algorithm)} - compressed file successfully: `);
  const keyLengths = Array.from(compressMap.keys(), (name) => name.length);
  const maxKeyLength = Math.max(...keyLengths);
  compressMap.forEach((value, name) => {
    let { size, oldSize, cname } = value;
    const rName = normalizePath(cname).replace(normalizePath(`${config.root}/${config.build.outDir}/`), "");
    const sizeStr = `${oldSize.toFixed(2)}kb / ${algorithm}: ${size.toFixed(2)}kb`;
    config.logger.info(chalk.dim(config.build.outDir + "/") + chalk.blueBright(rName) + " ".repeat(2 + maxKeyLength - name.length) + " " + chalk.dim(sizeStr));
  });
  config.logger.info("\n");
}
module.exports = exportFn;
var src_default = exportFn;
export {
  src_default as default
};
