// vite-plugin-vue-setup-extend/dist/index.mjs
// src/index.ts
import { parse, compileScript } from "@vue/compiler-sfc";
import MagicString from "magic-string";
var src_default = (options = {}) => {
  return {
    name: "vite:setup-name-support",
    enforce: "pre",
    async transform(code, id) {
      if (!/\.vue$/.test(id)) {
        return null;
      }
      const { name = true } = options;
      if (name) {
        return supportScriptName(code, id);
      }
    }
  };
};
function supportScriptName(code, id) {
  let s;
  const str = () => s || (s = new MagicString(code));
  const { descriptor } = parse(code);
  if (!descriptor.script && descriptor.scriptSetup) {
    const result = compileScript(descriptor, { id });
    const name = result.attrs.name;
    const lang = result.attrs.lang;
    if (name) {
      str().appendLeft(0, `
<script ${lang ? `lang="${lang}"` : ""}>
  import { defineComponent } from 'vue'
  export default defineComponent({
  name: '${name}',
})
<\/script>
`);
    }
    return {
      map: str().generateMap({ hires: true }),
      code: str().toString()
    };
  } else {
    return null;
  }
}
export {
  src_default as default
};
