// element-plus/es/components/teleport/src/teleport2.mjs
import { defineComponent, ref, computed, openBlock, createBlock, Teleport as Teleport$1, createElementVNode, normalizeStyle, renderSlot, createCommentVNode } from 'vue';
import { elTeleportProps } from './teleport.mjs';
import _export_sfc from '../../../_virtual/plugin-vue_export-helper.mjs';

const _sfc_main = defineComponent({
  props: elTeleportProps,
  setup(props) {
    const containerRef = ref();
    const containerStyle = computed(() => {
      return props.container === "body" ? [
        props.style,
        {
          position: "absolute",
          top: `0px`,
          left: `0px`,
          zIndex: props.zIndex
        }
      ] : {};
    });
    return {
      containerRef,
      containerStyle
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return _ctx.container ? (openBlock(), createBlock(Teleport$1, {
    key: 0,
    to: _ctx.container,
    disabled: _ctx.disabled
  }, [
    createElementVNode("div", {
      ref: "containerRef",
      class: "el-teleport",
      style: normalizeStyle(_ctx.containerStyle)
    }, [
      renderSlot(_ctx.$slots, "default")
    ], 4)
  ], 8, ["to", "disabled"])) : createCommentVNode("v-if", true);
}
var Teleport = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);

export { Teleport as default };
